R/data.table.R

```diff
@@ -2715,7 +2715,14 @@ setcolorder = function(x, neworder=key(x), before=NULL, after=NULL)  # before/af
 
 set = function(x,i=NULL,j,value)  # low overhead, loopable
 {
-  .Call(Cassign,x,i,j,NULL,value)
+  if (is.character(j)) {
+    name = substitute(x)
+    x = .Call(Cassign,x,i,j,NULL,value)
+    if (is.name(name))
+      assign(as.character(name),x,parent.frame(),inherits=TRUE)
```

Comment:
This `assign()` looks pretty complicated to reason about, I would like to see more tests of robustness here (`setDT()` within a function, places where `inherits=TRUE` is important, etc.).

At a glance I can't tell why this is needed though -- it definitely warrants a comment explaining why we need to branch.

Reply:
I simplified now the whole code. I also added a comment why we need the assign.

Ultimately it would be nice to let `set` work also in the cases where it needs to overallocate and is inside a function, howeover, for making this to work we would need to find the right environment where to assign (might not be the first where x is present) and it is not clear whether it would be the last due to scoping